artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPSessionCallback.java

```diff
@@ -543,7 +552,7 @@ private void inSessionSend(final ProtonServerReceiverContext context,
             afterIO(new IOCallback() {
                @Override
                public void done() {
-                  connection.runLater(() -> {
+                  connection.runNow(() -> {
```

Comment:
@tabish121 this will also fix your test, as it will avoid using an executor for non durable

Reply:
I tested this.. and the number of pending settlements is always 0 after this change for non durables... with durables it can go to a few dozen at a times.. .but credits will now take pending settlements into consideration.